Switch to Into<Option<Pid>> for Pid arguments in unistd

antifuchs opened the pull request:
This PR adjusts the unistd functions `setpgid`, `getpgid` and `getsid` to take `Into<Option<Pid>>` arguments, which (as of Rust 1.12) allows users to pass plain `Pid`s without wrapping them in `Some`.

Additionally, this allows users to pass `None` as both the `pid` and the `pgrp` argument in `setpgrp`, defaulting them to 0, as specified by POSIX.

I hope this is useful! If you like this, please let me know if I should go on an `Into<Option>` spree on the rest of the code (:

asomers commented:
LGTM, but it needs a CHANGELOG entry.

Susurrus commented on CHANGELOG.md:
```diff
@@ -37,6 +37,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
+- `setpgid`, `getpgid` and `getsid` now take `Into<Option<Pid>>`
+  arguments. It is now possible to specify these optional Pid
+  arguments as plain `Pid`s or as `None`.
```
Please link to the PR as the other CHANGELOG entries do.

Susurrus commented on src/unistd.rs:
```diff
@@ -240,20 +240,23 @@ pub fn getppid() -> Pid {
-/// may be used to group together a set of processes in order for the OS to
-/// apply some operations across the group.
+/// Set the process group id (PGID) of a particular process.  If None
+/// or a pid of zero is specified, then the pid of the calling process
```
Do we really want to still support a Pid of 0 here? I'd argue that the semantics would be improved by forcing people o use `None` or a valid `Pid` rather than having two possible values for `None`.

antifuchs replied:
I have spent a while thinking about this now, and I don't see a better option than the one here:

* To disallow `0`, we'd need to be able to have an `.into` that can fail. The only way I know to do that would be to use `TryInto`, which isn't stable yet.
* If we disallowed `0` pid arguments, we'd become incompatible with other code that treats `0` as the "self" pid argument (e.g. C code).
* It's a super incompatible change with rust client code, as stuff that passed in `0` pids before no longer can (the program for which I am making this change would be one of them).

So I'm not sure those tradeoffs are worth introducing the additional complexity (and ergonomic setbacks) of a pid-but-not-zero type. Does that make sense?

Susurrus commented on src/unistd.rs:
```diff
@@ -240,20 +240,23 @@ pub fn getppid() -> Pid {
 #[inline]
-pub fn getpgid(pid: Option<Pid>) -> Result<Pid> {
-    let res = unsafe { libc::getpgid(pid.unwrap_or(Pid(0)).into()) };
+pub fn getpgid<P: Into<Option<Pid>>>(pid: P) -> Result<Pid> {
```
Can you add doccomments for this function since you're modifying it?

Susurrus commented on src/unistd.rs:
```diff
@@ -268,14 +271,13 @@ pub fn setsid() -> Result<Pid> {
 ///
 /// Obtain the process group ID of the process that is the session leader of the process specified
-/// by pid. If pid is zero, it specifies the calling process.
+/// by pid. If passed None or pid is or zero, it specifies the calling process.
```
Same as the above, do we really want to allow `0` as a valid input. I'd argue that we shouldn't.

Susurrus commented on test/test_unistd.rs:
```diff
@@ -112,8 +112,10 @@ fn test_getpid() {
 fn test_getsid() {
     let none_sid: ::libc::pid_t = getsid(None).unwrap().into();
     let pid_sid: ::libc::pid_t = getsid(Some(getpid())).unwrap().into();
+    let plain_sid: ::libc::pid_t = getsid(getpid()).unwrap().into();
```
None of these test cases cover `0`. I'm a fan of removing that functionality, but wanted to point it out for completeness.

antifuchs commented:
I squashed, added the changelog link, documented `getpgid` and added the test case; as outlined above, I'm not sure a 0-disallowing interface is feasible/super desirable at the moment, but if you can think of a way to do it nicely, I'm happy to change it!
